### PR TITLE
registry/storage/cache/memory: Use LRU cache to bound cache size

### PR DESCRIPTION
**Pull request by aaronlehmann**
Instead of letting the cache grow without bound, use a LRU to impose a
size limit.

The limit is configurable through a new `blobdescriptorsize` config key.

### CODE REVIEW

File: registry/handlers/app.go [hunk 1]
```diff
@@ -280,7 +281,17 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
 			}
 			dcontext.GetLogger(app).Infof("using redis blob descriptor cache")
 		case "inmemory":
-			cacheProvider := memorycache.NewInMemoryBlobDescriptorCacheProvider()
+			blobDescriptorSize := 10000
```
Comment (thaJeztah) [inline registry/handlers/app.go:284]: Only glancing over; perhaps would be good to have  a `const` for this (and use that `const` to document what it's for, and where the `10000` came from (which may be just "we had to pick a value"). Such values can cause some head-scratching in future, where people forgot if there was a specific reason for picking the value.

File: registry/handlers/app_test.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ import (
 func TestAppDispatcher(t *testing.T) {
 	driver := testdriver.New()
 	ctx := context.Background()
-	registry, err := storage.NewRegistry(ctx, driver, storage.BlobDescriptorCacheProvider(memorycache.NewInMemoryBlobDescriptorCacheProvider()), storage.EnableDelete, storage.EnableRedirect)
+	registry, err := storage.NewRegistry(ctx, driver, storage.BlobDescriptorCacheProvider(memorycache.NewInMemoryBlobDescriptorCacheProvider(0)), storage.EnableDelete, storage.EnableRedirect)
```
Comment (thaJeztah) [inline registry/handlers/app_test.go:29]: Perhaps the same for the `0` (which means "unlimited?")

File: registry/handlers/app.go (hunk 1 above)
Comment (aaronlehmann) [reply registry/handlers/app.go]: Done

File: registry/handlers/app_test.go (hunk 2 above)
Comment (aaronlehmann) [reply registry/handlers/app_test.go:29]: Done

**Comment by aaronlehmann** [comment]
@milosgajdos: Any chance you could take a look at this when you have some time?

File: registry/storage/cache/memory/memory.go [hunk 3]
```diff
@@ -30,39 +50,59 @@ func (imbdcp *inMemoryBlobDescriptorCacheProvider) RepositoryScoped(repo string)
...
+	if err := dgst.Validate(); err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	key := descriptorCacheKey{
+		digest: dgst,
+	}
+	descriptor, ok := imbdcp.lru.Get(key)
+	if ok {
+		return descriptor.(distribution.Descriptor), nil
```
Comment (milosgajdos) [inline registry/storage/cache/memory/memory.go:69]: I _suspect_ skipping type assertion here will trigger OSS fuzzer (event though it does not make much sense here)

**Review by milosgajdos — COMMENTED** [review]
I like this and I think this can be valuable. The slight problem I have is introducing a config param in `cache` section that only applies to a specific type of cache -- users will inevitably start setting this on other types of cache (`redis`) and be asking questions why this isn't working and why are they redis instances running out of memory or whatnot (yes I can see the readme mentions it only applies to `inmemory` descriptor cache)

**Comment by aaronlehmann** [comment]
> The slight problem I have is introducing a config param in cache section that only applies to a specific type of cache -- users will inevitably start setting this on other types of cache (`redis`) and be asking questions why this isn't working and why are they redis instances running out of memory or whatnot (yes I can see the readme mentions it only applies to `inmemory` descriptor cache)

I'm totally open to moving or renaming the config parameter. What do you think makes most sense? I could rename it to `inmemorysize`, or add an `inmemory` section underneath `cache`.

File: registry/storage/cache/memory/memory.go (hunk 3 above)
Comment (aaronlehmann) [reply registry/storage/cache/memory/memory.go]: I've added type assertions just in case.

**Comment by milosgajdos** [comment]
> I'm totally open to moving or renaming the config parameter. What do you think makes most sense? I could rename it to inmemorysize, or add an inmemory section underneath cache.

I must say I **was** thinking about this -- I'm not entirely against leaving it as is, but that'd have that caveat I mentioned 😬 
Maybe @thaJeztah can weigh in here just so it's not just the two of us. I'll tag @deleteriousEffect on the review as well.

**Comment by thaJeztah** [comment]
Without diving into depth (so possibly silly questions); is there any chance that the option will be supported by non-memory-backed caches? (if so, then I don't think renaming would be good?)

But If if doesn't take effect on the configured driver, would it make sense to produce an error (for the user to discover the issue and to remove the option from their config)?

**Comment by aaronlehmann** [comment]
> Without diving into depth (so possibly silly questions); is there any chance that the option will be supported by non-memory-backed caches? (if so, then I don't think renaming would be good?)

I think the only other cache driver is redis, and for that one I believe it makes more sense to configure retention on the server side. I'm not very familiar with that cache driver so I'm not sure whether it's possible or desirable to support the option for that driver.

> But If if doesn't take effect on the configured driver, would it make sense to produce an error (for the user to discover the issue and to remove the option from their config)?

That sounds good to me.

**Comment by aaronlehmann** [comment]
I've added a warning when `blobdescriptorsize` is specified in combination with the Redis blob descriptor cache. This seems most consistent with the behavior when an unrecognized cache driver is specified (also a warning), but I can change this to a blocking error if you prefer.

**Comment by aaronlehmann** [comment]
@milosgajdos @thaJeztah: Are you happy with this approach to making `blobdescriptorsize` specific to `inmemory`?

**Comment by aaronlehmann** [comment]
Gentle ping on this one. I believe I've addressed all the outstanding feedback.

**Review by milosgajdos — APPROVED** [review]
I think this looks good to me 🤔 CC: @thaJeztah

**Comment by aaronlehmann** [comment]
@milosgajdos @thaJeztah: Anything else I can do to get this merged?

**Comment by milosgajdos** [comment]
I've ticked it ✅ It's with @thaJeztah now :)

**Review by thaJeztah — COMMENTED** [review]
Sorry for the delay (saw your ping, and of course then forgot about it)

changes look good to me, but given that the last three commits are touching-up changes added in the first commit, I'd prefer squashing them; could you squash them?

**Comment by aaronlehmann** [comment]
Rebased

**Review by thaJeztah — APPROVED** [review]
LGTM

@milosgajdos ptal

**Pull request merged**